File: vi/codes/java/chapter_computational_complexity/space_complexity.java
/**
 * File: space_complexity.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_computational_complexity;

import utils.*;
import java.util.*;

public class space_complexity {
    /* Hàm */
    static int function() {
        // Thực hiện một số thao tác
        return 0;
    }

    /* Độ phức tạp không gian hằng số */
    static void constant(int n) {
        // Hằng số, biến, đối tượng chiếm không gian O(1)
        final int a = 0;
        int b = 0;
        int[] nums = new int[10000];
        ListNode node = new ListNode(0);
        // Biến trong vòng lặp chiếm không gian O(1)
        for (int i = 0; i < n; i++) {
            int c = 0;
        }
        // Hàm trong vòng lặp chiếm không gian O(1)
        for (int i = 0; i < n; i++) {
            function();
        }
    }

    /* Độ phức tạp không gian tuyến tính */
    static void linear(int n) {
        // Mảng có độ dài n sử dụng không gian O(n)
        int[] nums = new int[n];
        // Danh sách có độ dài n chiếm không gian O(n)
        List<ListNode> nodes = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            nodes.add(new ListNode(i));
        }
        // Bảng băm có độ dài n chiếm không gian O(n)
        Map<Integer, String> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(i, String.valueOf(i));
        }
    }

    /* Độ phức tạp không gian tuyến tính (triển khai đệ quy) */
    static void linearRecur(int n) {
        System.out.println("Đệ quy n = " + n);
        if (n == 1)
            return;
        linearRecur(n - 1);
    }

    /* Độ phức tạp không gian bậc hai */
    static void quadratic(int n) {
        // Ma trận sử dụng không gian O(n^2)
        int[][] numMatrix = new int[n][n];
        // Danh sách 2 chiều sử dụng không gian O(n^2)
        List<List<Integer>> numList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            List<Integer> tmp = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                tmp.add(0);
            }
            numList.add(tmp);
        }
    }

    /* Độ phức tạp không gian bậc hai (triển khai đệ quy) */
    static int quadraticRecur(int n) {
        if (n <= 0)
            return 0;
        // Mảng nums có độ dài n, n-1, ..., 2, 1
        int[] nums = new int[n];
        System.out.println("Trong đệ quy n = " + n + ", độ dài nums = " + nums.length);
        return quadraticRecur(n - 1);
    }

    /* Mã điều khiển */
    static TreeNode buildTree(int n) {
        if (n == 0)
            return null;
        TreeNode root = new TreeNode(0);
        root.left = buildTree(n - 1);
        root.right = buildTree(n - 1);
        return root;
    }

    /* Mã điều khiển */
    public static void main(String[] args) {
        int n = 5;
        // Độ phức tạp không gian hằng số
        constant(n);
        // Độ phức tạp không gian tuyến tính
        linear(n);
        linearRecur(n);
        // Độ phức tạp không gian bậc hai
        quadratic(n);
        quadraticRecur(n);
        // Độ phức tạp không gian theo cấp số mũ
        TreeNode root = buildTree(n);
        PrintUtil.printTree(root);
    }
}

File: vi/codes/swift/chapter_computational_complexity/space_complexity.swift
/**
 * File: space_complexity.swift
 * Created Time: 2023-01-01
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Hàm */
@discardableResult
func function() -> Int {
    // Thực hiện một số thao tác
    return 0
}

/* Độ phức tạp không gian hằng số */
func constant(n: Int) {
    // Hằng số, biến, đối tượng chiếm không gian O(1)
    let a = 0
    var b = 0
    let nums = Array(repeating: 0, count: 10000)
    let node = ListNode(x: 0)
    // Biến trong vòng lặp chiếm không gian O(1)
    for _ in 0 ..< n {
        let c = 0
    }
    // Hàm trong vòng lặp chiếm không gian O(1)
    for _ in 0 ..< n {
        function()
    }
}

/* Độ phức tạp không gian tuyến tính */
func linear(n: Int) {
    // Mảng có độ dài n chiếm không gian O(n)
    let nums = Array(repeating: 0, count: n)
    // Danh sách liên kết có độ dài n chiếm không gian O(n)
    let nodes = (0 ..< n).map { ListNode(x: $0) }
    // Bảng băm có độ dài n chiếm không gian O(n)
    let map = Dictionary(uniqueKeysWithValues: (0 ..< n).map { ($0, "\($0)") })
}

/* Độ phức tạp không gian tuyến tính (triển khai đệ quy) */
func linearRecur(n: Int) {
    print("Recursion n = \(n)")
    if n == 1 {
        return
    }
    linearRecur(n: n - 1)
}

/* Độ phức tạp không gian theo cấp số nhân */
func quadratic(n: Int) {
    // Danh sách 2 chiều chiếm không gian O(n^2)
    let numList = Array(repeating: Array(repeating: 0, count: n), count: n)
}

/* Độ phức tạp không gian bậc hai (triển khai đệ quy) */
@discardableResult
func quadraticRecur(n: Int) -> Int {
    if n <= 0 {
        return 0
    }
    // Mảng nums có độ dài n, n-1, ..., 2, 1
    let nums = Array(repeating: 0, count: n)
    print("In recursion n = \(n), nums length = \(nums.count)")
    return quadraticRecur(n: n - 1)
}

/* Mã điều khiển */
func buildTree(n: Int) -> TreeNode? {
    if n == 0 {
        return nil
    }
    let root = TreeNode(x: 0)
    root.left = buildTree(n: n - 1)
    root.right = buildTree(n: n - 1)
    return root
}

@main
enum SpaceComplexity {
    /* Mã điều khiển */
    static func main() {
        let n = 5
        // Độ phức tạp không gian hằng số
        constant(n: n)
        // Độ phức tạp không gian tuyến tính
        linear(n: n)
        linearRecur(n: n)
        // Độ phức tạp không gian theo cấp số nhân
        quadratic(n: n)
        quadraticRecur(n: n)
        // Độ phức tạp không gian theo cấp số nhân
        let root = buildTree(n: n)
        PrintUtil.printTree(root: root)
    }
}
